PropertyChanged.Fody.Analyzer/SourceGeneratorEngine.cs

```diff
@@ -0,0 +1,85 @@
+﻿using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+#pragma warning disable CS0067
+static class SourceGeneratorEngine
+{
+    const string sourceFileHintName = "PropertyChanged.g.cs";
+
+    public static void GenerateSource(SourceProductionContext context, Compilation compilation, Configuration configuration, ImmutableArray<ClassDeclarationSyntax> classes)
+    {
+        var cancellationToken = context.CancellationToken;
+
+        if (configuration.IsDisabled)
+        {
+            context.AddSource(sourceFileHintName, @"// Source generator is disabled by configuration.");
+            return;
+        }
+
+        var eventInvokerName = configuration.EventInvokerName.NullIfEmpty() ?? "OnPropertyChanged";
+
+        var codeBuilder = new CodeBuilder();
+
+        try
+        {
+            codeBuilder
+                .Add("// <auto-generated/>")
+                .Add("#nullable enable")
+                .Add("#pragma warning disable CS0067")
+                .Add("using System.ComponentModel;")
+                .Add("using System.Runtime.CompilerServices;");
+
+            var classesByNamespace = classes.GroupBy(item => item.SyntaxTree)
```

Comment:
It can do both now, but reading through the documentation again I don't see the benefit of having one file per class, as the generator part will always be run for all candidates.
It just makes it harder to filter out duplicates.

Reply:
That was just a question, I was being curious. 🙂

I thought you could avoid having to use `Collect`, but as you say, you need to deduplicate.

Reply:
Me too, was an interesting deep dive into code generator details 😀